[JSC] Remove the redundant NodeResultInt32 check in addImmediateShouldSpeculateInt32

hyjorc1 opened the pull request:
#### fd6adbcc872d8be55be0968e34f10f14b9c19619
<pre>
[JSC] Remove the redundant NodeResultInt32 check in addImmediateShouldSpeculateInt32
<a href="https://bugs.webkit.org/show_bug.cgi?id=271322">https://bugs.webkit.org/show_bug.cgi?id=271322</a>
<a href="https://rdar.apple.com/125099867">rdar://125099867</a>

Reviewed by Keith Miller.

This path removes the redundant NodeResultInt32 check for
the non-integer constant double case at the end of
addImmediateShouldSpeculateInt32.

* Source/JavaScriptCore/dfg/DFGGraph.h:

Canonical link: <a href="https://commits.webkit.org/276983@main">https://commits.webkit.org/276983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/533082fc6cd0307d65c02e2eb07b7560524a71ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47484 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40835 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21324 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36818 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45408 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20986 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17891 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39740 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2877 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38026 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41061 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40031 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49148 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44286 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16352 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43807 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21116 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42563 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10253 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21455 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51458 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20791 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10443 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -1318,10 +1318,12 @@ class Graph final : public virtual Scannable {
         NodeFlags operandResultType = operand->result();
         if (operandResultType != NodeResultInt32 && immediateValue.isDouble())
             return DontSpeculateInt32;
-        
```
Not your issue but can you change the `Node*immediate` to `Node* immediate` above?

kmiller68 commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -1331,13 +1333,12 @@ class Graph final : public virtual Scannable {
         if (bytecodeCanTruncateInteger(add->arithNodeFlags())) {
-            // If int32 + const double, then we should not speculate this add node with int32 type.
-            // Because ToInt32(int32 + const double) is not always equivalent to int32 + ToInt32(const double).
+            // This function is used in attemptToMakeIntegerAdd. If succeed, both operands will be truncate to integers.
```
Nit: "This function is called from attemptToMakeIntegerAdd. If we return SpeculateInt32AndTruncateConstants both operands will be truncated to integers."

kmiller68 reviewed (approved):
r=me.

hyjorc1 committed:
[JSC] Remove the redundant NodeResultInt32 check in addImmediateShouldSpeculateInt32
https://bugs.webkit.org/show_bug.cgi?id=271322
rdar://125099867

Reviewed by Keith Miller.

This path removes the redundant NodeResultInt32 check for
the non-integer constant double case at the end of
addImmediateShouldSpeculateInt32.

* Source/JavaScriptCore/dfg/DFGGraph.h:

Canonical link: https://commits.webkit.org/276983@main
